L85-core-binder-transport.md

```diff
@@ -0,0 +1,181 @@
+called by Android, the channel will be created and the callback will be invoked.
+
+Because the API's signature contains [JNI] specific type `jobject`, the API will
+only be declared if the header is compiled with Android tool chain. (that is,
+when `GPR_ANDROID` is defined)
+
+### Server side API
+
+Service implementation will be an Android [bound service] in Java, and it should return
+a binder object created by our transport implementation to the client.
+
+We propose to introduce a new API that returns credentials for binder server:
+
+```Cpp
+std::shared_ptr<ServerCredentials> BinderServerCredentials(std::shared_ptr<grpc::binder::SecurityPolicy> security_policy);
+```
+
+The following snippet sets up gRPC server listening for incoming binder
+transactions, and only allows clients that signed by the same key to connect:
+
+```Cpp
+grpc::ServerBuilder server_builder;
+server_builder.AddListeningPort(
+    "binder://example",
+    grpc::BinderServerCredentials(grpc::binder::SameSignatureSecurityPolicy()));
+```
+
+The first argument specifies a "binder port". A binder port is not a real port,
+and is identified with a custom binder URI scheme `binder://`.
+
+After `server_builder.BuildAndStart()` is called, an endpoint binder will be
+created internally in the transport.
+
+To let the client connect to the server, the service needs to return the
+endpoint binder. To get the endpoint binder, we provide a Java class `GrpcCppServerBuilder` and a static Java method
+that can be used like the following:
+
+```java
+GrpcCppServerBuilder::GetEndpointBinder("example")
+```
+
+The string parameter corresponds to the first argument passed in
+`grpc::ServerBuilder::AddListeningPort`.
+
+Note that the native library containing our transport implementation must
+have been loaded by user before using this helper class.
+
+Under the hood we will use a shared static variable to store the mapping between
+service name and endpoint binder.
+
+## Notes
+
+### Why asynchronous callback is used in client side API
+
+Due to Android API's design, Android service might not call
+`onServiceConnected` before current activity ends.
+
+User should not block the (android) Activity to wait for the callback to be
+called. Instead, user should unblock the thread and wait Android to notify that
+the bind is successful, in an asynchronous manner.
+
+We made the design decision to not return a (pending) channel object immediately
+because typically user will want to be notified when the channel become usable, and
+we want to prevent user from using the (not connected) channel immediately and
+wait for RPC call to success and causing deadlock.
+
+### Parcelable objects
+
+The protocol and Java implementation allows passing Parcelable objects. We
+currently don't have a plan to support it in C++. When we see a parcelable
+object in `Parcel`, we will print an error message and fail the RPC call.
+
+### Error handling
```

Comment:
Could you also make a note about what happens to the BinderChannel if the remote process is killed? Will the channel automatically try to reconnect the way a TCP or UDP channel does, or does the calling code need to manage reconnection?

Reply:
Added. Our Java class will receive notification when the connection is broken, and our code should be able to handle the connection recovery (and check authorization again if the server app is updated) for user.

Ref: https://developer.android.com/reference/android/content/ServiceConnection